content/docs/instrumenting/http_sd.md

```diff
@@ -0,0 +1,96 @@
+---
+title: HTTP SD
+sort_rank: 7
+---
+
+# Writing HTTP Service Discovery
+
+Prometheus provides a generic [HTTP Service Discovery](https://prometheus.io/docs/prometheus/latest/configuration/configuration/#http_sd_config),
+that enables it to discover targets over an HTTP endpoint.
+
+The HTTP Service Discovery is complimentary to the supported service
+discovery mechanisms, and is an alternative to [File-based Service Discovery](https://prometheus.io/docs/guides/file-sd/#use-file-based-service-discovery-to-discover-scrape-targets).
+
+## Comparison between File-Based SD and HTTP SD
+
+Here is a table comparing our two generic Service Discovery implementations.
+
+| Item | File SD | HTTP SD |
+| ---- | ------- | ------- |
+| Event Based | Yes, via inotify | No |
+| Update frequency | Instant, thanks to inotify | Following refresh_interval |
+| Format | Yaml or JSON | JSON |
+| Transport | Local file | HTTP/HTTPS |
+| Security | File-Based security | TLS, Basic auth, Authorization header, OAuth2 |
+
+## Requirements of HTTP SD endpoints
+
+If you implement an HTTP SD endpoint, here is a few requirements you should be
+aware of.
+
+The response is consumed as is, unmodified. On each refresh interval (default: 1
+minute), Prometheus will do a GET request to the HTTP SD endpoint. The GET
+request contains a `X-Prometheus-Refresh-Interval-Seconds` HTTP header with the
+refresh interval.
+
+The SD endpoint must answer with an HTTP 200 response, with the HTTP Header
+`Content-Type: application/json`. The answer must be UTF-8 formatted.
+If no targets should be transmitted, HTTP 200 must also be emitted, with
+an empty list `[]`.
+
+Prometheus caches target lists. If an error occurs while fetching an updated
+targets list, Prometheus keeps using the current targets list. The targets list
+is not saved across restart.
+
+The whole list of targets must be returned on every scrape. There is no support
+for incremental updates. A Prometheus instance does not send its hostname and it
+is not possible for a SD endpoint to know if the SD requests is the first one
+after a restart or not.
+
```

Comment:
random (probably irrelevant) question, need there be any consideration for sorting the results of the HTTP endpoint? would it cause unnecessary updates to the list of targets (potentially resetting scrape timers or something unintended) if the results "changed" by being in a different order?